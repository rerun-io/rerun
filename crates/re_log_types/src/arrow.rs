use std::collections::BTreeMap;

use arrow2::{
    array::{Array, PrimitiveArray},
    datatypes::{DataType, Field, TimeUnit},
};

use crate::{TimePoint, TimeType};

pub const OBJPATH_KEY: &str = "RERUN:object_path";
pub const TIMELINE_KEY: &str = "RERUN:timeline";
pub const TIMELINE_SEQUENCE: &str = "Sequence";
pub const TIMELINE_TIME: &str = "Time";

/// Build a iterator of (field, col) for all timelines in `time_point`
pub fn build_time_cols(
    time_point: &TimePoint,
) -> impl Iterator<Item = (Field, Box<dyn Array>)> + '_ {
    time_point.0.iter().map(|(timeline, time)| {
        let (datatype, meta_value) = match timeline.typ() {
            TimeType::Sequence => (DataType::Int64, TIMELINE_SEQUENCE),
            TimeType::Time => (
                DataType::Timestamp(TimeUnit::Nanosecond, None),
                TIMELINE_TIME,
            ),
        };
        let arr = PrimitiveArray::from([Some(time.as_i64())]).to(datatype);
        let field =
            Field::new(timeline.name().as_str(), arr.data_type().clone(), false).with_metadata(
                BTreeMap::from([(TIMELINE_KEY.to_owned(), meta_value.to_owned())]),
            );
        (field, arr.boxed())
    })
}

pub fn filter_time_cols<'a>(
    fields: &'a [Field],
    cols: &'a [Box<dyn Array>],
) -> impl Iterator<Item = (&'a Field, &'a Box<dyn Array>)> {
    fields
        .iter()
        .zip(cols.iter())
        .filter(|(field, _)| field.metadata.contains_key(TIMELINE_KEY))
}
