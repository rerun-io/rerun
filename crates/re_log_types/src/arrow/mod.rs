pub mod field_types;
mod time;
pub use time::*;

use arrow2::{array::MutableStructArray, datatypes::DataType};

use arrow2_convert::{field::ArrowField, serialize::ArrowSerialize};

pub mod exports {
    pub use arrow2_convert::serialize::TryIntoArrow;
}

use crate::{MsgId, ObjPath, Time, TimePoint, TimeType, Timeline};

pub struct ArrowLogMsg<C>
where
    C: ArrowField,
{
    msg_id: MsgId,
    time_point: TimePoint,
    object_path: ObjPath,
    components: C,
}

impl<C> ArrowField for ArrowLogMsg<C>
where
    C: ArrowField,
{
    type Type = Self;

    fn data_type() -> DataType {
        DataType::Extension(
            "ArrowLogMsg".to_owned(),
            Box::new(DataType::Struct(vec![
                <re_tuid::Tuid as ArrowField>::field("msg_id"),
                <TimePoint as ArrowField>::field("time_point"),
                <String as ArrowField>::field("object_path"),
                //<C as ArrowField>::field("components"),
            ])),
            None,
        )
    }
}

impl<C> ArrowSerialize for ArrowLogMsg<C>
where
    C: ArrowField<Type = C> + ArrowSerialize + 'static,
{
    type MutableArrayType = MutableStructArray;

    fn new_array() -> Self::MutableArrayType {
        let msg_id = Box::new(<re_tuid::Tuid as ArrowSerialize>::new_array());
        let time_point = Box::new(<TimePoint as ArrowSerialize>::new_array());
        let object_path = Box::new(<String as ArrowSerialize>::new_array());
        //let components = Box::new(<C as ArrowSerialize>::new_array());
        MutableStructArray::new(
            Self::data_type(),
            vec![
                msg_id,
                time_point,
                object_path,
                //components
            ],
        )
    }

    fn arrow_serialize(
        v: &<Self as ArrowField>::Type,
        array: &mut Self::MutableArrayType,
    ) -> arrow2::error::Result<()> {
        <re_tuid::Tuid as ArrowSerialize>::arrow_serialize(&v.msg_id.0, array.value(0).unwrap())?;
        <TimePoint as ArrowSerialize>::arrow_serialize(&v.time_point, array.value(1).unwrap())?;
        <String as ArrowSerialize>::arrow_serialize(
            &v.object_path.to_string(),
            array.value(2).unwrap(),
        )?;
        //<C as ArrowSerialize>::arrow_serialize(&v.components, array.value(3).unwrap())?;
        Ok(())
    }
}

#[test]
fn test_arrow_log_msg() {
    use arrow2::array::Array;
    use exports::TryIntoArrow;

    let mut time_point = TimePoint::default();
    time_point.0.insert(
        Timeline::new("log_time", TimeType::Time),
        Time::now().into(),
    );

    let msgs = [
        ArrowLogMsg {
            msg_id: MsgId::random(),
            time_point: time_point.clone(),
            object_path: ObjPath::from("obj1"),
            components: field_types::Point2D { x: 0.0, y: 1.0 },
        },
        ArrowLogMsg {
            msg_id: MsgId::random(),
            time_point: time_point.clone(),
            object_path: ObjPath::from("obj1"),
            components: field_types::Point2D { x: 1.0, y: 0.0 },
        },
    ];

    let array: Box<dyn Array> = msgs.try_into_arrow().unwrap();
    println!("{:#?}", array);
}

pub mod util {
    use std::collections::BTreeMap;

    use arrow2::{
        array::{Array, ListArray, MutableArray, PrimitiveArray, StructArray},
        buffer::Buffer,
        chunk::Chunk,
        datatypes::{DataType, Field, Schema, TimeUnit},
        error::Error,
    };
    use arrow2_convert::serialize::ArrowSerialize;

    use crate::{MsgId, TimePoint, TimeType};

pub type ComponentName = String;
pub type ComponentNameRef<'a> = &'a str;

pub const ENTITY_PATH_KEY: &str = "RERUN:entity_path";
    pub const TIMELINE_KEY: &str = "RERUN:timeline";
    pub const TIMELINE_SEQUENCE: &str = "Sequence";
    pub const TIMELINE_TIME: &str = "Time";

    pub fn build_log_msg_array(
        msg_id: MsgId,
        time_point: &TimePoint,
        components: StructArray,
    ) -> Result<(Chunk<Box<dyn Array>>, Schema), arrow2::error::Error> {
        let msg_id = {
            let mut m = re_tuid::Tuid::new_array();
            re_tuid::Tuid::arrow_serialize(&msg_id.0, &mut m)?;
            m.as_box()
        };

        let timelines = {
            // Build columns for timeline data
            let (fields, cols): (Vec<Field>, Vec<Box<dyn Array>>) =
                build_time_cols(time_point).unzip();
            StructArray::try_new(DataType::Struct(fields), cols, None)?
        };

        let schema = Schema::from(vec![
            Field::new("msg_id", msg_id.data_type().clone(), false),
            Field::new("timelines", timelines.data_type().clone(), false),
            Field::new("components", components.data_type().clone(), false),
        ]);

        let chunk = Chunk::try_new(vec![msg_id, timelines.boxed(), components.boxed()])?;

        Ok((chunk, schema))
    }

    /// Build a iterator of (field, col) for all timelines in `time_point`
    pub fn build_time_cols(
        time_point: &TimePoint,
    ) -> impl Iterator<Item = (Field, Box<dyn Array>)> + '_ {
        time_point.0.iter().map(|(timeline, time)| {
            let (datatype, meta_value) = match timeline.typ() {
                TimeType::Sequence => (DataType::Int64, TIMELINE_SEQUENCE),
                TimeType::Time => (
                    DataType::Timestamp(TimeUnit::Nanosecond, None),
                    TIMELINE_TIME,
                ),
            };
            let arr = PrimitiveArray::from([Some(time.as_i64())]).to(datatype);
            let field =
                Field::new(timeline.name().as_str(), arr.data_type().clone(), false).with_metadata(
                    BTreeMap::from([(TIMELINE_KEY.to_owned(), meta_value.to_owned())]),
                );
            (field, arr.boxed())
        })
    }

    pub fn filter_time_cols<'a>(
        fields: &'a [Field],
        cols: &'a [Box<dyn Array>],
    ) -> impl Iterator<Item = (&'a Field, &'a Box<dyn Array>)> {
        fields
            .iter()
            .zip(cols.iter())
            .filter(|(field, _)| field.metadata.contains_key(TIMELINE_KEY))
    }

    /// Wrap `field_array` in a single-element `ListArray`
    pub fn wrap_in_listarray(field_array: Box<dyn Array>) -> Result<ListArray<i32>, Error> {
        let datatype = ListArray::<i32>::default_datatype(field_array.data_type().clone());
        let offsets = Buffer::from(vec![0, field_array.len() as i32]);
        let values = field_array;
        let validity = None;
        ListArray::<i32>::try_new(datatype, offsets, values, validity)
    }
}
