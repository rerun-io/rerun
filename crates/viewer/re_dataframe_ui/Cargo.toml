[package]
authors.workspace = true
description = "Rich table widget over DataFusion."
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "re_dataframe_ui"
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true
include.workspace = true

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true

[dependencies]
re_arrow_util.workspace = true
re_chunk_store.workspace = true
re_dataframe.workspace = true
re_log_types.workspace = true
re_sorbet.workspace = true
re_tracing.workspace = true
re_types.workspace = true
re_types_core.workspace = true
re_ui.workspace = true
re_uri.workspace = true
re_viewer_context.workspace = true


ahash.workspace = true
arrow.workspace = true
datafusion.workspace = true
egui.workspace = true
egui_dnd.workspace = true
egui_table.workspace = true
nohash-hasher.workspace = true
parking_lot.workspace = true
serde.workspace = true
thiserror.workspace = true


[dev-dependencies]
static_assertions.workspace = true
