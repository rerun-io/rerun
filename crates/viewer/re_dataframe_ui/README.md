# re_dataframe_ui

Part of the [`rerun`](https://github.com/rerun-io/rerun) family of crates.

[![Latest version](https://img.shields.io/crates/v/re_dataframe_ui.svg)](https://crates.io/crates/re_dataframe_ui)
[![Documentation](https://docs.rs/re_dataframe_ui/badge.svg)](https://docs.rs/re_dataframe_ui)
![MIT](https://img.shields.io/badge/license-MIT-blue.svg)
![Apache](https://img.shields.io/badge/license-Apache-blue.svg)

Rich table widget over DataFusion.
