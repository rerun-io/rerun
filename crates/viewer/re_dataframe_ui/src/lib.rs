//! Rich table widget over `datafusion`.

mod datafusion_adapter;
mod datafusion_table_widget;
mod display_record_batch;
mod requested_object;
mod table_blueprint;
pub mod table_utils;

pub use datafusion_table_widget::DataFusionTableWidget;
pub use display_record_batch::{DisplayRecordBatch, DisplayRecordBatchError};
pub use requested_object::RequestedObject;
