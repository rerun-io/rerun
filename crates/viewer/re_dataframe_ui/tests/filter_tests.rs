use std::str::FromStr as _;
use std::sync::Arc;

use arrow::array::{
    Array, ArrayRef, BooleanArray, ListArray, ListBuilder, PrimitiveArray, PrimitiveBuilder,
    StringArray,
};
use arrow::buffer::{NullBuffer, OffsetBuffer};
use arrow::compute::cast;
use arrow::datatypes::{
    ArrowPrimitiveType, DataType, Field, Float32Type, Float64Type, Int8Type, Int16Type, Int32Type,
    Int64Type, Schema, TimeUnit, TimestampNanosecondType, UInt8Type, UInt16Type, UInt32Type,
    UInt64Type,
};
use arrow::record_batch::RecordBatch;
use datafusion::catalog::MemTable;
use datafusion::prelude::{DataFrame, SessionContext};
use jiff::ToSpan as _;

use re_dataframe_ui::{
    ComparisonOperator, Filter, FilterKind, FloatFilter, IntFilter, NonNullableBooleanFilter,
    Nullability, NullableBooleanFilter, TimestampFilter,
};
use re_viewer_context::external::tokio;

const COLUMN_NAME: &str = "column";
const SOME_TIMESTAMP: &str = "2025-09-23T11:47Z";

/// A single column of test data, with convenient constructors.
#[derive(Debug, Clone)]
struct TestColumn {
    field: Field,
    array: ArrayRef,
}

impl TestColumn {
    fn new(array: impl Array + 'static, nullable: bool) -> Self {
        let array = Arc::new(array) as ArrayRef;
        let field = Field::new(COLUMN_NAME, array.data_type().clone(), nullable);

        Self { array, field }
    }

    /// Create a primitive array with the provided data.
    fn primitive<T>(data: Vec<T::Native>) -> Self
    where
        T: ArrowPrimitiveType,
        PrimitiveArray<T>: From<Vec<T::Native>>,
    {
        Self::new(PrimitiveArray::<T>::from(data), false)
    }

    /// Create a nullable primitive array with the provided data.
    fn primitive_nulls<T>(data: Vec<Option<T::Native>>) -> Self
    where
        T: ArrowPrimitiveType,
        PrimitiveArray<T>: From<Vec<Option<T::Native>>>,
    {
        Self::new(PrimitiveArray::<T>::from(data), true)
    }

    /// Create a list array with the provided data and automatically injecting nulls as required.
    fn primitive_lists<T>(data: &[Vec<T::Native>], nullability: Nullability) -> Self
    where
        T: ArrowPrimitiveType,
        T::Native: Copy,
    {
        assert!(!data.is_empty());

        let value_builder = PrimitiveBuilder::<T>::new();
        let mut builder = ListBuilder::new(value_builder);

        for item in data {
            for inner_item in item {
                builder.values().append_value(*inner_item);
            }
            builder.append(true);
        }

        // inject outer null
        if nullability.outer {
            builder.append(false);
        }

        // inject inner nulls
        if nullability.inner {
            // lone null
            builder.values().append_null();
            builder.append(true);

            // find a non-empty inner item
            let non_empty_item = data
                .iter()
                .find(|item| !item.is_empty())
                .expect("there should be at least some non-empty data");

            // first null
            builder.values().append_null();
            for inner_item in non_empty_item {
                builder.values().append_value(*inner_item);
            }
            builder.append(true);

            // last null
            for inner_item in non_empty_item {
                builder.values().append_value(*inner_item);
            }
            builder.values().append_null();
            builder.append(true);
        }

        let array = builder.finish();

        Self::new(array, nullability.outer)
    }

    fn strings() -> Self {
        Self::new(
            StringArray::from(vec!["a", "b", "c", "ab", "A B", "aBc"]),
            false,
        )
    }

    fn strings_nulls() -> Self {
        Self::new(
            StringArray::from(vec![
                Some("a"),
                Some("b"),
                None,
                Some("ab"),
                Some("A B"),
                Some("aBc"),
            ]),
            true,
        )
    }

    fn strings_lists(nullability: Nullability) -> Self {
        // the primitive array stuff doesn't work for strings, so we go the manual way.
        let values = if nullability.inner {
            StringArray::from(vec![
                Some("a"),
                Some("b"),
                None,
                Some("ab"),
                None,
                Some("aBc"),
            ])
        } else {
            StringArray::from(vec!["a", "b", "c", "ab", "A B", "aBc"])
        };
        let offsets = OffsetBuffer::new(vec![0i32, 2, 4, 6].into());
        let strings_lists = ListArray::try_new(
            Arc::new(Field::new("item", DataType::Utf8, nullability.inner)),
            offsets,
            Arc::new(values),
            nullability
                .outer
                .then(|| NullBuffer::from(vec![true, false, true])),
        )
        .expect("failed to create a string list array");

        Self::new(strings_lists, nullability.outer)
    }

    fn bools() -> Self {
        Self::new(BooleanArray::from(vec![true, true, false]), false)
    }

    fn bools_nulls() -> Self {
        Self::new(
            BooleanArray::from(vec![Some(true), Some(true), None, Some(false)]),
            true,
        )
    }

    fn bool_lists(nullability: Nullability) -> Self {
        // the primitive array stuff doesn't work for bools, so we go the manual way.
        let values = if nullability.inner {
            BooleanArray::from(vec![
                Some(true),
                Some(false),
                None,
                Some(true),
                Some(false),
                None,
                Some(true),
                Some(false),
            ])
        } else {
            BooleanArray::from(vec![true, false, true, true, false, false, true, false])
        };

        let offsets = OffsetBuffer::new(vec![0i32, 1, 2, 4, 6, 8].into());
        let bool_lists = ListArray::try_new(
            Arc::new(Field::new("item", DataType::Boolean, nullability.inner)),
            offsets,
            Arc::new(values),
            nullability
                .outer
                .then(|| NullBuffer::from(vec![true, false, true, true, true])),
        )
        .expect("failed to create a bool list array");

        Self::new(bool_lists, nullability.outer)
    }

    fn timestamps(unit: TimeUnit) -> Self {
        let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

        let nano_column = Self::primitive::<TimestampNanosecondType>(vec![
            timestamp_to_nanos(some_date),
            timestamp_to_nanos(some_date - 1.hours()),
            timestamp_to_nanos(some_date - 24.hours()),
            timestamp_to_nanos(some_date - 8760.hours()),
            timestamp_to_nanos(some_date + 1.hours()),
            timestamp_to_nanos(some_date + 24.hours()),
            timestamp_to_nanos(some_date + 8760.hours()),
        ]);

        convert_timestamp_column(nano_column, unit)
    }

    fn timestamps_nulls(unit: TimeUnit) -> Self {
        let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

        let nano_column = Self::primitive_nulls::<TimestampNanosecondType>(vec![
            Some(timestamp_to_nanos(some_date)),
            Some(timestamp_to_nanos(some_date - 1.hours())),
            Some(timestamp_to_nanos(some_date - 24.hours())),
            Some(timestamp_to_nanos(some_date - 8760.hours())),
            None,
            Some(timestamp_to_nanos(some_date + 1.hours())),
            Some(timestamp_to_nanos(some_date + 24.hours())),
            None,
            Some(timestamp_to_nanos(some_date + 8760.hours())),
        ]);

        convert_timestamp_column(nano_column, unit)
    }

    fn timestamps_lists(unit: TimeUnit, nullability: Nullability) -> Self {
        let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

        let nano_column = Self::primitive_lists::<TimestampNanosecondType>(
            &[
                vec![],
                vec![timestamp_to_nanos(some_date)],
                vec![
                    timestamp_to_nanos(some_date - 1.hours()),
                    timestamp_to_nanos(some_date - 24.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date),
                    timestamp_to_nanos(some_date - 168.hours()),
                    timestamp_to_nanos(some_date - 8760.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date + 1.hours()),
                    timestamp_to_nanos(some_date + 24.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date),
                    timestamp_to_nanos(some_date + 168.hours()),
                    timestamp_to_nanos(some_date + 8760.hours()),
                ],
                vec![
                    timestamp_to_nanos(some_date),
                    timestamp_to_nanos(some_date - 8760.hours()),
                    timestamp_to_nanos(some_date + 8760.hours()),
                ],
            ],
            nullability,
        );

        convert_timestamp_column(nano_column, unit)
    }
}

fn timestamp_to_nanos(ts: jiff::Timestamp) -> i64 {
    ts.as_nanosecond()
        .try_into()
        .expect("timestamp is too large")
}

fn convert_timestamp_column(nano_column: TestColumn, unit: TimeUnit) -> TestColumn {
    if unit == TimeUnit::Nanosecond {
        nano_column
    } else {
        TestColumn::new(
            cast(nano_column.array.as_ref(), &DataType::Timestamp(unit, None))
                .expect("timestamp column cast failed"),
            nano_column.field.is_nullable(),
        )
    }
}

/// A temporary session context populated with a "test" dataframe and constructed from a bunch
/// of test columns.
struct TestSessionContext {
    ctx: SessionContext,
}

impl TestSessionContext {
    fn new(columns: impl IntoIterator<Item = TestColumn>) -> Self {
        let ctx = SessionContext::new();

        let (fields, arrays): (Vec<_>, Vec<_>) =
            columns.into_iter().map(|c| (c.field, c.array)).unzip();

        let schema = Arc::new(Schema::new_with_metadata(fields, Default::default()));

        ctx.register_table(
            "__test",
            Arc::new(
                MemTable::try_new(
                    Arc::clone(&schema),
                    vec![vec![
                        RecordBatch::try_new_with_options(schema, arrays, &Default::default())
                            .expect("failed to create the record batch"),
                    ]],
                )
                .expect("failed to create mem table"),
            ),
        )
        .expect("failed to register table");

        Self { ctx }
    }

    async fn df(&self) -> DataFrame {
        self.ctx
            .table("__test")
            .await
            .expect("test table not found")
    }

    async fn to_filtered_record_batch(&self, filter: &Filter) -> RecordBatch {
        let df = self.df().await;

        let schema = df.schema();
        let filter_expr = filter
            .as_filter_expression(schema)
            .expect("couldn't create an expression from filter");

        let mut record_batches = df
            .filter(filter_expr)
            .expect("failed to apply filter")
            .collect()
            .await
            .expect("failed to collect");

        assert_eq!(record_batches.len(), 1);

        record_batches
            .pop()
            .expect("we just checked that there is one record batch")
    }
}

#[derive(Debug)]
#[expect(dead_code)] // debug is excluded from dead code analysis
struct TestResult<'a> {
    op: FilterKind,
    field: Field,
    unfiltered: ArrayRef,
    filtered: &'a ArrayRef,
}

//note: this is a macro so insta is exposed to the actual test function
macro_rules! filter_snapshot {
    ($filter_op:expr, $test_column:expr, $case:expr) => {
        let filter = Filter::new(COLUMN_NAME, $filter_op.clone());
        let initial_field = $test_column.field.clone();
        let initial_column = $test_column.array.clone();

        let result = TestSessionContext::new([$test_column])
            .to_filtered_record_batch(&filter)
            .await;

        assert_eq!(result.columns().len(), 1);
        assert_eq!(&initial_field, result.schema().field(0));

        let final_column = result.column(0);


        let test_results = TestResult {
            op: $filter_op,
            field: initial_field,
            unfiltered: initial_column,
            filtered: final_column,
        };

        insta::with_settings!({
           snapshot_suffix => $case,
        },
        {
            insta::assert_debug_snapshot!(test_results);
        });
    };
}

#[tokio::test]
async fn test_int_compares() {
    let ints = TestColumn::primitive::<Int64Type>(vec![1, 2, 3, 4, 5]);
    let ints_nulls =
        TestColumn::primitive_nulls::<Int64Type>(vec![Some(1), Some(2), None, Some(4), Some(5)]);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(3))),
            ints.clone(),
            format!("{}_3", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(4))),
            ints_nulls.clone(),
            format!("nulls_{}_4", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, None)),
            ints_nulls.clone(),
            format!("nulls_{}_unspecified", op.as_ascii())
        );
    }
}

/// Make sure we correctly handle all integer types.
#[tokio::test]
async fn test_int_all_types() {
    macro_rules! test_int_all_types_impl {
        ($ty:tt) => {
            filter_snapshot!(
                FilterKind::Int(IntFilter::new(ComparisonOperator::Eq, Some(3))),
                TestColumn::primitive::<$ty>(vec![1, 2, 3, 4, 5]),
                format!("{:?}", $ty {})
            )
        };
    }

    test_int_all_types_impl!(Int8Type);
    test_int_all_types_impl!(Int16Type);
    test_int_all_types_impl!(Int32Type);
    test_int_all_types_impl!(Int64Type);
    test_int_all_types_impl!(UInt8Type);
    test_int_all_types_impl!(UInt16Type);
    test_int_all_types_impl!(UInt32Type);
    test_int_all_types_impl!(UInt64Type);
}

#[tokio::test]
async fn test_int_lists() {
    let data = vec![
        vec![1, 2, 3],
        vec![2],
        vec![2, 2],
        vec![4, 5, 6],
        vec![7, 4, 9],
        vec![5, 2, 1],
    ];
    let int_lists = TestColumn::primitive_lists::<Int64Type>(&data, Nullability::NONE);
    let int_lists_nulls = TestColumn::primitive_lists::<Int64Type>(&data, Nullability::BOTH);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(2))),
            int_lists.clone(),
            format!("{}_2", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Int(IntFilter::new(*op, Some(2))),
            int_lists_nulls.clone(),
            format!("nulls_{}_2", op.as_ascii())
        );
    }
}

#[tokio::test]
async fn test_float_compares() {
    let floats = TestColumn::primitive::<Float64Type>(vec![1.0, 2.0, 3.0, 4.0, 5.0]);
    let floats_nulls = TestColumn::primitive_nulls::<Float64Type>(vec![
        Some(1.0),
        Some(2.0),
        None,
        Some(4.0),
        Some(5.0),
    ]);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(3.0))),
            floats.clone(),
            format!("{}_3.0", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(4.0))),
            floats_nulls.clone(),
            format!("nulls_{}_4", op.as_ascii())
        );
    }
}

/// Make sure we correctly handle all float types.
#[tokio::test]
async fn test_float_all_types() {
    macro_rules! test_float_all_types_impl {
        ($ty:tt) => {
            filter_snapshot!(
                FilterKind::Float(FloatFilter::new(ComparisonOperator::Eq, Some(3.0))),
                TestColumn::primitive::<$ty>(vec![1.0, 2.0, 3.0, 4.0, 5.0]),
                format!("{:?}", $ty {})
            )
        };
    }

    test_float_all_types_impl!(Float32Type);
    test_float_all_types_impl!(Float64Type);
}

#[tokio::test]
async fn test_float_lists() {
    let data = vec![
        vec![1.0, 2.0, 3.0],
        vec![2.0],
        vec![2.0, 2.0],
        vec![4.0, 5.0, 6.0],
        vec![7.0, 4.0, 9.0],
        vec![5.0, 2.0, 1.0],
    ];
    let float_lists = TestColumn::primitive_lists::<Float64Type>(&data, Nullability::NONE);
    let float_lists_nulls = TestColumn::primitive_lists::<Float64Type>(&data, Nullability::BOTH);

    for op in ComparisonOperator::ALL {
        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(2.0))),
            float_lists.clone(),
            format!("{}_2.0", op.as_ascii())
        );

        filter_snapshot!(
            FilterKind::Float(FloatFilter::new(*op, Some(2.0))),
            float_lists_nulls.clone(),
            format!("nulls_{}_2.0", op.as_ascii())
        );
    }
}

#[tokio::test]
async fn test_string_contains() {
    filter_snapshot!(
        FilterKind::StringContains(String::new()),
        TestColumn::strings(),
        "empty"
    );

    filter_snapshot!(
        FilterKind::StringContains("a".to_owned()),
        TestColumn::strings(),
        "a"
    );

    filter_snapshot!(
        FilterKind::StringContains("a".to_owned()),
        TestColumn::strings(),
        "ab"
    );

    filter_snapshot!(
        FilterKind::StringContains("A".to_owned()),
        TestColumn::strings(),
        "a_uppercase"
    );

    filter_snapshot!(
        FilterKind::StringContains(String::new()),
        TestColumn::strings_nulls(),
        "nulls_empty"
    );

    filter_snapshot!(
        FilterKind::StringContains("a".to_owned()),
        TestColumn::strings_nulls(),
        "nulls_a"
    );
}

#[tokio::test]
async fn test_string_contains_list() {
    for &nullability in Nullability::ALL {
        filter_snapshot!(
            FilterKind::StringContains("ab".to_owned()),
            TestColumn::strings_lists(nullability),
            format!("{nullability:?}_ab")
        );
    }
}

/// Non-nullable filter should work regardless of nullability.
#[tokio::test]
async fn test_non_nullable_boolean_equals() {
    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsTrue),
        TestColumn::bools(),
        "true"
    );

    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsFalse),
        TestColumn::bools(),
        "false"
    );

    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsTrue),
        TestColumn::bools_nulls(),
        "nulls_true"
    );

    filter_snapshot!(
        FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsFalse),
        TestColumn::bools_nulls(),
        "nulls_false"
    );
}

#[tokio::test]
async fn test_nullable_boolean_equals() {
    filter_snapshot!(
        FilterKind::NullableBoolean(NullableBooleanFilter::IsTrue),
        TestColumn::bools_nulls(),
        "nulls_true"
    );

    filter_snapshot!(
        FilterKind::NullableBoolean(NullableBooleanFilter::IsFalse),
        TestColumn::bools_nulls(),
        "nulls_false"
    );

    filter_snapshot!(
        FilterKind::NullableBoolean(NullableBooleanFilter::IsNull),
        TestColumn::bools_nulls(),
        "nulls_null"
    );
}

#[tokio::test]
async fn test_boolean_equals_list_non_nullable() {
    for &nullability in Nullability::ALL {
        filter_snapshot!(
            FilterKind::NonNullableBoolean(NonNullableBooleanFilter::IsTrue),
            TestColumn::bool_lists(nullability),
            format!("{nullability:?}_true")
        );
    }
}

#[tokio::test]
async fn test_boolean_equals_list_nullable() {
    // Note: NullableBooleanFilter doesn't support Nullability::NONE, but that's ok because
    // NonNullableBooleanFilter is used in this case.
    for nullability in [Nullability::BOTH, Nullability::INNER, Nullability::OUTER] {
        filter_snapshot!(
            FilterKind::NullableBoolean(NullableBooleanFilter::IsNull),
            TestColumn::bool_lists(nullability),
            format!("{nullability:?}")
        );
    }
}

const ALL_TIME_UNITS: &[TimeUnit] = &[
    TimeUnit::Second,
    TimeUnit::Millisecond,
    TimeUnit::Microsecond,
    TimeUnit::Nanosecond,
];

#[tokio::test]
async fn test_timestamps() {
    // Note: this test intends to cover all column datatypes. It does not intend to cover all kinds
    // of timestamp filtering, which is already covered by unit tests.

    let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

    let all_filters = [
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date)),
            "after",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date + 1.seconds())),
            "after_strict",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::between(
                some_date - 168.hours(),
                some_date - 2.hours(),
            )),
            "between",
        ),
    ];

    for time_unit in ALL_TIME_UNITS {
        for (filter, case) in &all_filters {
            for nullable in [true, false] {
                filter_snapshot!(
                    filter.clone(),
                    if nullable {
                        TestColumn::timestamps_nulls(*time_unit)
                    } else {
                        TestColumn::timestamps(*time_unit)
                    },
                    format!(
                        "{case}_{time_unit:?}{}",
                        if nullable { "_nulls" } else { "" }
                    )
                );
            }
        }
    }
}

#[tokio::test]
async fn test_timestamps_list() {
    // Note: this test intends to cover all column datatypes. It does not intend to cover all kinds
    // of timestamp filtering, which is already covered by unit tests.

    let some_date = jiff::Timestamp::from_str(SOME_TIMESTAMP).expect("valid");

    let all_filters = [
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date)),
            "after",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::after(some_date + 1.seconds())),
            "after_strict",
        ),
        (
            FilterKind::Timestamp(TimestampFilter::between(
                some_date - 168.hours(),
                some_date - 2.hours(),
            )),
            "between",
        ),
    ];

    for (filter, case) in &all_filters {
        for &nullability in Nullability::ALL {
            filter_snapshot!(
                filter.clone(),
                TestColumn::timestamps_lists(TimeUnit::Nanosecond, nullability),
                format!("{case}_{nullability:?}")
            );
        }
    }
}
