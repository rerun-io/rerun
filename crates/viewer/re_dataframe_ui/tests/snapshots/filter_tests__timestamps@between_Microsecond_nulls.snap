---
source: crates/viewer/re_dataframe_ui/tests/filter_tests.rs
expression: test_results
---
TestResult {
    op: Timestamp(
        TimestampFilter {
            kind: Between,
            low_bound_timestamp: EditableTimestamp {
                timestamp_string: None,
                resolved_timestamp: Ok(
                    2025-09-16T11:47:00Z,
                ),
            },
            high_bound_timestamp: EditableTimestamp {
                timestamp_string: None,
                resolved_timestamp: Ok(
                    2025-09-23T09:47:00Z,
                ),
            },
        },
    ),
    field: Field {
        name: "column",
        data_type: Timestamp(
            Microsecond,
            None,
        ),
        nullable: true,
        dict_id: 0,
        dict_is_ordered: false,
        metadata: {},
    },
    unfiltered: PrimitiveArray<Timestamp(Microsecond, None)>
    [
      2025-09-23T11:47:00,
      2025-09-23T10:47:00,
      2025-09-22T11:47:00,
      2024-09-23T11:47:00,
      null,
      2025-09-23T12:47:00,
      2025-09-24T11:47:00,
      null,
      2026-09-23T11:47:00,
    ],
    filtered: PrimitiveArray<Timestamp(Microsecond, None)>
    [
      2025-09-22T11:47:00,
    ],
}
